onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -31,58 +31,59 @@ namespace onnxruntime {
+  // Get shape and type of the tensor, and allocate the empty tensor
+  const auto& dims = tensor_proto.dims();
+  std::vector<int64_t> tensor_shape_vec(static_cast<size_t>(dims.size()));
+  for (int i = 0; i < dims.size(); ++i) {
+    tensor_shape_vec[i] = dims[i];
   }
-
-  // deserialize to CPU first for non-CPU allocator, then alloc and copy
-  size_t cpu_tensor_length;
-  ORT_RETURN_IF_ERROR(utils::GetSizeInBytesFromTensorProto<0>(tensor_proto, &cpu_tensor_length));
-  if (m.GetLen() < cpu_tensor_length) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Internal error. The preallocated buffer is too small. Requires ",
-                           cpu_tensor_length, ", Got ", m.GetLen());
+  TensorShape tensor_shape{tensor_shape_vec};
+  const DataTypeImpl* const type = DataTypeImpl::TensorTypeFromONNXEnum(tensor_proto.data_type())->GetElementType();
+  std::unique_ptr<Tensor> p_tensor;
+  if (m != nullptr) {
+    p_tensor = onnxruntime::make_unique<Tensor>(type, tensor_shape, m->GetBuffer(), m->GetAllocInfo());
+    if (m->GetLen() < p_tensor->SizeInBytes()) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Internal error. The preallocated buffer is too small. Requires ",
+                             p_tensor->SizeInBytes(), ", Got ", m->GetLen());
+    }
+  } else {
+    // tensor constructor should give us enough buffer size based on type and shape
+    p_tensor = onnxruntime::make_unique<Tensor>(type, tensor_shape, alloc);
   }
 
-  std::unique_ptr<char[]> data(new char[cpu_tensor_length]);
-  std::unique_ptr<Tensor> p_tensor;
-  OrtValue tmp_ort_value;
-  OrtCallback d;
-  ORT_RETURN_IF_ERROR(utils::TensorProtoToMLValue(env, proto_path.c_str(), tensor_proto,
-                                                  MemBuffer(data.get(), cpu_tensor_length, default_cpu_memory_info),
-                                                  tmp_ort_value, d));
-
-  const Tensor& p_deserialize_tensor = tmp_ort_value.Get<Tensor>();
-
-  p_tensor = onnxruntime::make_unique<Tensor>(p_deserialize_tensor.DataType(), p_deserialize_tensor.Shape(), m.GetBuffer(),
-                                              m.GetAllocInfo());
-  // TODO: does this function work for string tensor?
-  Status copy_status = data_transfer_mgr.CopyTensor(p_deserialize_tensor, *p_tensor);
-  if (d.f) d.f(d.param);
-
-  if (!copy_status.IsOK()) {
-    if (copy_status.ErrorMessage().empty()) {
-      // The windows execution provider does not return any error message today for CopyTensor since it is
-      // not implemented yet. That's the reason we're adding our own error message so that we can debug better.
-      return Status(copy_status.Category(), copy_status.Code(),
-                    "Failed to copy tensor to " + p_tensor->Location().ToString());
+  if (strcmp(p_tensor->Location().name, CPU) == 0 || p_tensor->Location().mem_type == OrtMemTypeCPUOutput) {
+    // deserialize directly to CPU tensor
```

Comment:
This is just deserializing initializers isn't it? If so, there's no need to check for something like OrtMemTypeCPUOutput as that only happens during graph execution